### PR TITLE
Enforce Operation Requirements

### PR DESCRIPTION
**Pull request by b-butler**
## Description

This enforces two primary constraints we have in our work flow model:
1. Operations should not be condition functions and vice versa
2. A function and/or name should only be used once for all operations (not just the name-function pair)
This shouldn't be considered an API break in my opinion, but if someone is writing unorthodox scripts this could technically break their work-flow.

## Motivation and Context

These changes are in line with our current API design.

## Types of Changes

- [ ] Documentation update
- [x] Bug fix
- [ ] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

**Review by bdice — APPROVED** [review]
For the historical record, this PR was split off from #487. LGTM.

**Comment by bdice** [comment]
@b-butler It would be good to add a changelog line. Here are some cute dinosaurs who say thank you for the PR. 🦖 🦕

**Comment by b-butler** [comment]
@bdice thanks dinos! Also I had the changelog entry, but I forgot to push it. :rofl:

File: changelog.txt [hunk 1]
```diff
@@ -26,6 +26,8 @@ Changed
 Fixed
 +++++
 
+- Enforce operations condition separation and single use of names and functions
+as operations (#496).
```
Comment (csadorf) [inline changelog.txt:30]: there should be no line break here

File: flow/project.py [hunk 2]
```diff
@@ -4089,11 +4089,24 @@ def hello(job):
         if isinstance(func, str):
             return lambda op: cls.operation(op, name=func)
 
+        if func in chain(
+            *cls._OPERATION_PRECONDITIONS.values(),
+            *cls._OPERATION_POSTCONDITIONS.values(),
+        ):
+            raise ValueError("An condition function cannot be used as an operation")
```
Comment (csadorf) [inline flow/project.py:4096]:
```suggestion
            raise ValueError("An condition function cannot be used as an operation.")
```

File: changelog.txt [hunk 3]
```diff
@@ -26,6 +26,7 @@ Changed
 Fixed
 +++++
 
+- Enforce operations condition separation and single use of names and functions as operations (#496).
```
Comment (csadorf) [inline changelog.txt:29]: I  find the changelog entry a little cryptic. How about this:
```suggestion
- Strictly enforce that operation functions cannot be used as condition functions (and vice-versa) and prevent the registration of two operations with the same name. (#496).
```

**Review by csadorf — APPROVED** [review]
Two suggestions, but overall good. Thank you very much @b-butler !

**Pull request merged**